### ISSUE TITLE
measure_acceptance scores the draft against penalised target logits: a draft identical to the target scores 0.86

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#345]
Found while validating #303 on an H100.

`measure_acceptance` calls `generate(do_sample=False)`, which disables **sampling** but not the checkpoint's `repetition_penalty` logits processor. The target then emits `argmax(penalised logits)` while the draft is scored on `argmax(raw logits)` — so the two are compared under different rules.

**A draft identical to the target does not score 1.0:**

| self-acceptance (draft == target) | declared rep-penalty | shipped path | penalty neutralised |
|---|---|---|---|
| Qwen2.5-0.5B | 1.1 | **0.8623** | **0.9892** |
| Llama-3.2-1B | 1.0 | 0.9945 | 0.9945 |
| Llama-3.1-8B | 1.0 | 1.000000 | 1.000000 |

**Numerical precision was the obvious hypothesis and it was tested and REJECTED:** fp32 gives 0.8638 against bf16's 0.8623, and the median top-2 logit gap at disagreement is 0.64–0.75 — decisive margins, not near-ties.

**This does not exist in real speculative decoding.** Confirmed from source that assisted generation passes the *same* `logits_processor` to the draft (`candidate_generator.py:314`). So the defect is in the measurement, not in the feature it measures — which is worse in one specific way: it misreports the thing users gate on.

**Practical impact:** on Qwen2.5-32B it **flips the verdict band** — 0.6857 MODERATE becomes 0.7072 STRONG once the penalty is neutralised. A CI gate at `--min-acceptance 0.70` would fail that pair for an artefact.

Models that declare no `repetition_penalty` are unaffected, so previously published SmolLM2 figures stand.

### Also found in the same subsystem
- **`--steps N` delivers N/4.44 steps.** `_build_distill_config_yaml` computes `epochs = ceil(steps / data_rows)` ignoring `val_split=0.1` and `gradient_accumulation_steps=4`; `--steps 2000` over 400 rows ran **450** optimizer steps. Epochs over the data are as intended; the user-facing knob is not.
- **The teacher loads in fp32** (`distill.py` passes no dtype) — 32 GB for an 8B teacher, ~131 GB for the 32B one #303 proposes. Same root cause as #339.
- `soup draft distill` captures its `soup train` subprocess output entirely, so there is no live progress or loss curve.

### Acceptance
- [ ] the draft is scored under the same logits processors the target used
- [ ] a draft identical to the target scores 1.0 (the control that would have caught this)
- [ ] `--steps N` produces N optimizer steps, or the flag is renamed to what it does

Record: `benchmarks/gate-h100-validation.md` STEP 26.

**Comment by MakazhanAlpamys** [comment]
Closing — all three acceptance boxes are satisfied on `main` today. Re-verified each one against the code rather than against a commit message, and the one loose end is tracked elsewhere (below).

**1. "the draft is scored under the same logits processors the target used."** `utils/draft.py:551` pins the target's generate call to `"repetition_penalty": 1.0`, commented `# neutralise checkpoint penalty: raw argmax agreement`. That is the "penalty neutralised" column from the table in this issue — both sides now compare on raw argmax, which is the same rule, applied symmetrically.

**2. "a draft identical to the target scores 1.0 (the control that would have caught this)."** `tests/test_v07133.py:1075`, `test_repetition_penalty_neutralized_for_self_acceptance`, docstring `(Refs #345)`. It drives an identical draft through `measure_acceptance` and asserts all three of the things worth asserting, not just the score:

```python
assert target.calls[0]["repetition_penalty"] == 1.0
assert total == 3
assert accepted == 3
```

Passes locally: `1 passed, 135 deselected`.

**3. "`--steps N` produces N optimizer steps, or the flag is renamed."** Fixed by arithmetic, not by rename. `commands/draft.py:267` `_distill_epochs_for_steps` now takes `val_split`, `batch_size` and `grad_accum` explicitly, and delegates the per-epoch count to `_distill_steps_per_epoch` (`:254`); the two defaults this issue identified as missing are named constants at `:74-75` (`_DISTILL_VAL_SPLIT = 0.1`, `_DISTILL_GRAD_ACCUM = 4`) instead of being implicit. Its docstring names this issue's number directly:

> Rounding up means the request is met or overshot by less than one epoch, never the ~1/4.44 undershoot of the old arithmetic.

Epoch granularity means `--steps` can still only land *near* the request, which is inherent to there being no `max_steps` knob in the trainer — and `:420` now prints the resolved epoch/accum/batch so any residual mismatch is visible to the operator rather than silent. That is the honest version of criterion 3.

**What is NOT fixed, and why it does not hold this issue open.** The "Also found in the same subsystem" note about the teacher loading in fp32 is still true — `trainer/distill.py:282` calls `AutoModelForCausalLM.from_pretrained` with `trust_remote_code` and `device_map` but no dtype. This issue said it itself: *"Same root cause as #339."* #339 is open with #471 in review against it, so the fix belongs there, where it gets applied to every load path at once instead of only the distill teacher. Tracking it in two places would just mean fixing it in one and forgetting the other.

The third note (`soup draft distill` capturing its subprocess output, so no live progress or loss curve) is a UX gap rather than a correctness defect and was never in the acceptance list. Worth its own issue if it still bites; not a reason to keep a fixed correctness bug open.
